src/inmanta/deploy/state.py

```diff
@@ -38,24 +38,36 @@ def requires_view(self) -> Mapping[ResourceIdStr, Set[ResourceIdStr]]:
     def provides_view(self) -> Mapping[ResourceIdStr, Set[ResourceIdStr]]:
         return self.reverse_mapping()
 
-    def get_all_provides_transitively(self, resource: ResourceIdStr | set[ResourceIdStr]) -> list[ResourceIdStr]:
+    def get_all_provides_transitively(self, resource: ResourceIdStr | Set[ResourceIdStr]) -> list[ResourceIdStr]:
         """
-        This method returns all the provides (transitively) of the given resource.
+        This method returns all the provides (transitively) of the given resource. The returned list will be sorted as such
```

Comment:
What is the motivation for sorting them? The current approach for sorting them seems like it may do a lot of double (or triple, quadruple, ...) work if the provides tree is heavily non-treelike.

Reply:
> What is the motivation for sorting them?

I see now that if we don't, it would complicate the `unblock_resources` logic because we would need to traverse the requires relation more deeply. i.e. what I said there about checking "if `r` is in `self.requires.get_all_provides_transitively(resource)`" doesn't suffice because `r` might have other blocked requires outside of that provides forest.

Reply:
With this in mind, I wonder if it might not be simpler to not use the method and simply traverse the relation directly, unblocking resources along the way and stopping recursion when either a resource is already blocked (will have already informed its provides) or when one of its requires is still blocked (this one won't be unblocked and neither will its children, unless we reach this node again).